// eas-cli/build/commandUtils/flags.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EasUpdateEnvironmentFlag = exports.EasJsonOnlyFlag = exports.EASNonInteractiveFlag = exports.EASEnvironmentVariableScopeFlag = exports.EASVariableVisibilityFlag = exports.EASVariableFormatFlag = exports.EASMultiEnvironmentFlag = exports.EASEnvironmentFlag = exports.EasEnvironmentFlagParameters = exports.EasNonInteractiveAndJsonFlags = void 0;
const core_1 = require("@oclif/core");
exports.EasNonInteractiveAndJsonFlags = {
    json: core_1.Flags.boolean({
        description: 'Enable JSON output, non-JSON messages will be printed to stderr.',
        dependsOn: ['non-interactive'],
    }),
    'non-interactive': core_1.Flags.boolean({
        description: 'Run the command in non-interactive mode.',
    }),
};
exports.EasEnvironmentFlagParameters = {
    description: "Environment variable's environment, e.g. 'production', 'preview', 'development'",
};
exports.EASEnvironmentFlag = {
    environment: core_1.Flags.string({
        description: "Environment variable's environment, e.g. 'production', 'preview', 'development'",
    }),
};
exports.EASMultiEnvironmentFlag = {
    environment: core_1.Flags.string({
        ...exports.EasEnvironmentFlagParameters,
        multiple: true,
    }),
};
exports.EASVariableFormatFlag = {
    format: core_1.Flags.enum({
        description: 'Output format',
        options: ['long', 'short'],
        default: 'short',
    }),
};
exports.EASVariableVisibilityFlag = {
    visibility: core_1.Flags.enum({
        description: 'Visibility of the variable',
        options: ['plaintext', 'sensitive', 'secret'],
    }),
};
exports.EASEnvironmentVariableScopeFlag = {
    scope: core_1.Flags.enum({
        description: 'Scope for the variable',
        options: ['project', 'account'],
        default: 'project',
    }),
};
exports.EASNonInteractiveFlag = {
    'non-interactive': core_1.Flags.boolean({
        description: 'Run the command in non-interactive mode.',
    }),
};
exports.EasJsonOnlyFlag = {
    json: core_1.Flags.boolean({
        description: 'Enable JSON output, non-JSON messages will be printed to stderr.',
    }),
};
exports.EasUpdateEnvironmentFlag = {
    environment: core_1.Flags.string({
        description: 'Environment to use for the server-side defined EAS environment variables during command execution, e.g. "production", "preview", "development"',
        required: false,
        default: undefined,
    }),
};
